helix-core/src/main/java/org/apache/helix/controller/stages/CustomizedViewAggregationStage.java

```diff
@@ -0,0 +1,171 @@
+
+    if (manager == null || resourceMap == null || cache == null) {
+      throw new StageException(
+          "Missing attributes in event:" + event + ". Requires ClusterManager|RESOURCES|DataCache");
+    }
+
+    HelixDataAccessor dataAccessor = manager.getHelixDataAccessor();
+    PropertyKey.Builder keyBuilder = dataAccessor.keyBuilder();
+
+    CustomizedStateOutput customizedStateOutput =
+        event.getAttribute(AttributeName.CUSTOMIZED_STATE.name());
+    ClusterStatusMonitor clusterStatusMonitor =
+        event.getAttribute(AttributeName.clusterStatusMonitor.name());
+
+    List<CustomizedView> newCustomizedViews = new ArrayList<>();
+    Set<String> monitoringResources = new HashSet<>();
+
+    Map<String, CustomizedViewCache> customizedViewCacheMap = cache.getCustomizedViewCacheMap();
+    cache.refreshCustomizedViewMap(dataAccessor);
+
+    // remove stale customized view type from ZK and cache
+    List<String> customizedViewTypesToRemove = new ArrayList<>();
+    for (String stateType : customizedViewCacheMap.keySet()) {
+      if (!customizedStateOutput.getAllStateTypes().contains(stateType)) {
+        LogUtil.logInfo(LOG, _eventId, "Remove customizedView for stateType: " + stateType);
+        dataAccessor.removeProperty(keyBuilder.customizedView(stateType));
+        customizedViewTypesToRemove.add(stateType);
+      }
+    }
+
+    cache.removeCustomizedViewTypes(customizedViewTypesToRemove);
+
+    // update customized view
+    for (String stateType : customizedStateOutput.getAllStateTypes()) {
+      Map<String, CustomizedView> curCustomizedViews = new HashMap<>();
+      CustomizedViewCache customizedViewCache = customizedViewCacheMap.get(stateType);
+      if (customizedViewCache != null) {
+        curCustomizedViews = customizedViewCache.getCustomizedViewMap();
+      } else {
+        customizedViewCache = new CustomizedViewCache(event.getClusterName(), stateType);
+      }
+
+      for (Resource resource : resourceMap.values()) {
+        try {
+          computeCustomizedStateView(resource, stateType, customizedStateOutput, curCustomizedViews,
+              newCustomizedViews);
+          // Keep MBeans for existing resources and unregister MBeans for dropped resources
+          if (clusterStatusMonitor != null) {
+            clusterStatusMonitor.retainResourceMonitor(monitoringResources);
+          }
+
+          List<PropertyKey> keys = new ArrayList<>();
+          for (Iterator<CustomizedView> it = newCustomizedViews.iterator(); it.hasNext(); ) {
+            CustomizedView view = it.next();
+            String resourceName = view.getResourceName();
+            keys.add(keyBuilder.customizedView(stateType, resourceName));
+          }
+          // add/update customized-views
+          if (newCustomizedViews.size() > 0) {
+            dataAccessor.setChildren(keys, newCustomizedViews);
+            customizedViewCache.refresh(dataAccessor);
```

Comment:
Shall we just pass the in-memory content to the cache instead of requiring it to read from ZK again?

Reply:
I tried to use in-memory content, and that does require a couple of set functions in CustomizedViewCache. I feel it's not clean. Please let me know if you think it's worth.

Reply:
I've confirmed that routing table would not read from the customized view cache output from this stage, so we are safe to remove the refresh of cache